File: typescript/api/src/gameTemplate.ts
import {FinalRoom} from "./rooms/finalRoom";
import {StartRoom} from "./rooms/startRoom";
import {GameOverRoom} from "./rooms/gameOverRoom";
import {Room} from "./base/room";

/**
 * Class that holds data related to the game design (all rooms, that first room ect)
 *
 */
export class GameTemplate {
    private static instance: GameTemplate;

    //All room in the game
    private _rooms: Room[] = [];

    //The room the gamer is in when the game starts
    //Should also be in rooms
    private _startRoom: Room = new StartRoom();

    private constructor() {
        this._rooms.push(this._startRoom);
        this._rooms.push(new FinalRoom());
        this._rooms.push(new GameOverRoom());
    }

    public getRoom(roomId: string): Room {
        const room: Room | undefined = this._rooms.find(function (room) {
            return room.id === roomId;
        });
        if (room) {
            return room;
        }
        console.log("Error room not found");
        return this._startRoom;
    }

    /**
     * Static method that returns the one and only instance (Singleton)
     *
     */
    public static getInstance(): GameTemplate {
        if (this.instance === undefined) {
            this.instance = new GameTemplate();
        }
        return this.instance;
    }

}

File: java/api/src/main/java/game/engine/java/api/GameTemplate.java
package game.engine.java.api;

import game.engine.java.api.base.Room;
import game.engine.java.api.rooms.BedRoom;
import game.engine.java.api.rooms.StartRoom;

import java.util.ArrayList;
import java.util.List;

/**
 * Class that holds data related to the game design (all rooms, that first room ect)
 */
public class GameTemplate {
    private static GameTemplate instance;
    private List<Room> rooms = new ArrayList();
    private Room startRoom = new StartRoom();
    private Room BedRoom = new BedRoom();

    private GameTemplate() {
        // TODO add other rooms -----
        this.rooms.add(this.BedRoom);
        this.rooms.add(this.startRoom);
    }

    /**
     * Static method that returns the one and only instance (Singleton)
     */
    public static GameTemplate getInstance() {
        if (GameTemplate.instance == null) {
            GameTemplate.instance = new GameTemplate();
        }
        return GameTemplate.instance;
    }

    // TODO return the room in this.rooms with matching id instead of the startRoom -----
    public Room getRoomById(String id) {
        for (Room room : this.rooms) {
            if (room.getId().equals(id)) {
                return room;
            }
        }
        return null;
    }

    public List<Room> getRooms() {
        return rooms;
    }

    public void setRooms(List<Room> rooms) {
        this.rooms = rooms;
    }

    public Room getStartRoom() {
        return startRoom;
    }

    public void setStartRoom(Room startRoom) {
        this.startRoom = startRoom;
    }

    public Room getBedRoom() {
        return BedRoom;
    }

    public void setBedRoom(Room BedRoom) {
        this.BedRoom = BedRoom;
    }


}
